### PR TITLE
[AssetCondition] Add asset_condition to AutoMaterializePolicy

### PR DESCRIPTION
**Pull request by OwenKephart**
## Summary & Motivation

This allows users to experiment with AssetCondition objects in shapes that are currently not possible with the AutoMaterializeRule system.

In the future, there will be a dedicated argument on the `@asset` decorator and others, but for now this provides a simple entry point.

### CODE REVIEW

**Comment by OwenKephart** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/19747?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `master`
  * **PR #19553** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/19553?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #19746** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/19746?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #19747** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/19747?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/19747?utm_source=stack-comment).

File: python_modules/dagster/dagster_tests/definitions_tests/auto_materialize_tests/updated_scenarios/custom_condition_scenarios.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+from .asset_daemon_scenario_states import (
+    one_asset,
+)
+
+custom_condition_scenarios = [
+    AssetDaemonScenario(
+        id="lotta_nots_scenario",
+        initial_state=one_asset.with_asset_properties(
+            auto_materialize_policy=AutoMaterializePolicy.from_asset_condition(
+                ~~~~~~~~~~RuleCondition(AutoMaterializeRule.materialize_on_missing())
```
Comment (alangenfeld) [inline python_modules/dagster/dagster_tests/definitions_tests/auto_materialize_tests/updated_scenarios/custom_condition_scenarios.py:18]: `~~~~~~~~~~` ?

Comment (OwenKephart) [reply python_modules/dagster/dagster_tests/definitions_tests/auto_materialize_tests/updated_scenarios/custom_condition_scenarios.py:18]: Ok that could have been explained better haha -- it's basically just doing a bunch of `NotAssetCondition(NotAssetCondition(...))` just as an example of a thing that has a different structure than is currently allowed by the AutoMaterializeRule system. I'll update this test.

Comment (alangenfeld) [reply python_modules/dagster/dagster_tests/definitions_tests/auto_materialize_tests/updated_scenarios/custom_condition_scenarios.py:18]: got it, comment is probably sufficient (but necessary)

**Review by alangenfeld — APPROVED** [review]

**Pull request merged**